### ISSUE TITLE
Scala 3

### ISSUE THREAD

**Issue by mkurz** [playframework/playframework#11260]
All issues and pull requests for Scala 3 support should be labeled with:
- https://github.com/playframework/playframework/labels/scala3

In preparation of adding support for Scala 3 to Playframework, it is helpful to tackle the support issue on a per-project basis and in an optimal order. The reason is that the PlayFramework build consists of a large number of (sub-)projects where a large number of them depend on other projects in the same build.

The idea is to apply the process project by project at consecutive 'levels'.

The following sections shows 13 levels, starting at level 0 (L0). When all projects at a certain level have been tackled, the next level is 'unlocked' and the projects in it can be tackled. Note that for each project, its dependencies at the project level are shown (` xxx -> yyy, zzz, ...`) for ease of reference.

```[tasklist]
### L0
- [x] ~SbtRoutesCompilerProject~ - not applicable
- [x] RoutesCompilerProject - #11302
- [x] ~PlayExceptionsProject~ - not applicable
- [x] StreamsProject - #11304 
- [ ] https://github.com/playframework/playframework/pull/11336
```

```[tasklist]
### L1
- [x] ~BuildLinkProject~ - not applicable
```

```[tasklist]
### L2
- [x] PlayProject - #11310
- [x] RunSupportProject
- [x] https://github.com/playframework/playframework/pull/11645
- [x] https://github.com/playframework/playframework/pull/11326
- [x] https://github.com/playframework/playframework/pull/11327
- [x] https://github.com/playframework/playframework/pull/11512
- [x] https://github.com/playframework/playframework/pull/11550
- [x] https://github.com/playframework/playframework/pull/11646
```

```[tasklist]
### L3
- [x] PlayJdbcApiProject          -> PlayProject
- [x] SbtPluginProject            -> RunSupportProject
- [x] PlayClusterSharding         -> PlayProject
- [x] PlayJavaClusterSharding     -> PlayProject
- [x] PlayGuiceProject            -> PlayProject
```

```[tasklist]
### L4
- [x] PlayDocsSbtPlugin           -> SbtPluginProject
- [x] SbtScriptedToolsProject     -> SbtPluginProject
- [x] PlayServerProject           -> PlayProject, PlayGuiceProject
```

```[tasklist]
### L5
- [x] PlayNettyServerProject      -> PlayServerProject
- [x] PlayAkkaHttpServerProject   -> PlayServerProject, PlayGuiceProject
```

```[tasklist]
### L6
- [x] PlayTestProject             -> PlayAkkaHttpServerProject, PlayServerProject, PlayGuiceProject
- [x] PlayAkkaHttp2SupportProject -> PlayAkkaHttpServerProject
- [x] PlayDocsProject             -> PlayAkkaHttpServerProject
```

```[tasklist]
### L7
- [x] PlaySpecs2Project           -> PlayTestProject
- [x] PlayWsProject               -> PlayTestProject, PlayProject
```

```[tasklist]
### L8
- [x] PlayJodaFormsProject        -> PlaySpecs2Project, PlayProject
- [x] PlayJavaProject             -> PlaySpecs2Project, PlayProject, PlayTestProject, PlayGuiceProject
- [x] PlayLogback                 -> PlaySpecs2Project, PlayProject
- [x] PlayCacheProject            -> PlaySpecs2Project, PlayProject
- [x] PlayJdbcProject             -> PlaySpecs2Project, PlayJdbcApiProject
```

```[tasklist]
### L9
- [x] PlayJavaJdbcProject         -> PlayJdbcProject, PlayJavaProject, PlayGuiceProject, PlaySpecs2Project
- [x] PlayJavaFormsProject        -> PlayJavaProject
- [x] PlayEhcacheProject          -> PlayCacheProject, PlaySpecs2Project, PlayProject
- [x] PlayCaffeineCacheProject    -> PlayCacheProject, PlaySpecs2Project, PlayProject
```

```[tasklist]
### L10
- [x] PlayJdbcEvolutionsProject   -> PlayJavaJdbcProject, PlayJdbcApiProject, PlayJdbcProject, PlaySpecs2Project
- [x] PlayJCacheProject           -> PlayCaffeineCacheProject, PlaySpecs2Project, PlayProject
- [x] PlayAhcWsProject            -> PlayCaffeineCacheProject, PlayWsProject, PlaySpecs2Project, PlayTestProject,PlayAkkaHttpServerProject
```

```[tasklist]
### L11
- [x] PlayOpenIdProject           -> PlayAhcWsProject, PlaySpecs2Project
- [x] PlayJpaProject              -> PlayJdbcEvolutionsProject, PlayJavaJdbcProject, PlaySpecs2Project
- [x] PlayFiltersHelpersProject   -> PlayAkkaHttpServerProject, PlayAhcWsProject, PlaySpecs2Project, PlayJavaProject, PlayTestProject, PlayProject
```

```[tasklist]
### L12
- [x] PlayIntegrationTestProject  -> PlayFiltersHelpersProject, PlayJavaProject, PlayJavaFormsProject, PlayAkkaHttpServerProject, PlayAkkaHttp2SupportProject, PlayNettyServerProject, PlaySpecs2Project, PlayProject, PlayLogback, PlayAhcWsProject, PlayServerProject
```

```[tasklist]
### L13
- [x] PlayMicrobenchmarkProject   -> PlayAhcWsProject, PlaySpecs2Project, PlayFiltersHelpersProject, PlayJavaProject, PlayNettyServerProject, PlayAkkaHttpServerProject, PlayAkkaHttp2SupportProject, PlayProject, PlayLogback, PlayIntegrationTestProject
```

```[tasklist]
### Notable issues from other repos
- [x] https://github.com/playframework/play-file-watch/issues/200
```

**Comment by xuwei-k** [comment]
I think done `L0` and `L1` (only compile. except test)

- Routes-Compiler, Play-Streams => Done
  - https://github.com/playframework/playframework/pull/11302
  - https://github.com/playframework/playframework/pull/11304
- Sbt-Routes-Compiler, RunSupportProject => unnecessary. this is Scala 2.12 (sbt) project
- Play-Exceptions, Build-Link => unnecessary. this is only java sources

next is `L2` the `Play` project( `core/play/src/main` )

There are some issues

- [ ] https://github.com/lampepfl/dotty/issues/12840
- [ ] translate Scala 2 macros
- [ ] `AkkaStreams.bypassWith` overload is ambiguous. I think we should add another name method
- [ ] other trivial changes?

**Comment by octonato** [comment]
Thanks @xuwei-k. 
I updated the description to reflect the progress.

**Comment by xuwei-k** [comment]
![graph](https://user-images.githubusercontent.com/389787/173283933-269fe925-88b5-4a68-9c72-0d797ef6b753.svg)

**Comment by mkurz** [comment]
So... actually I think it's finally time to tackle Scala 3 once more... I am currently reviewing some more pull requests but want to come back to this one asap. Would be nice if people can help to get this one going.

**Comment by mkurz** [comment]
It seems quite some people watch this tracking issue, so I want to give a quick update regarding Scala 3:

There are two pull requests
* #11550
* #11551 (based on top of the first one)

which make Play cross compile to Scala 3, publish Scala 3 artifacts locally and make all test pass in Scala 3 (and 2.13), including even scripted tests. This is very promising progress and I hope to merge that PR soon after some clean up and probably slight changes still to incorporate since the PRs are not perfect yet.

Shoutout to @jtjeferreira who started those pull requests!

**Comment by mkurz** [comment]
So... yesterday I tagged Play 2.9.0-M3 which now publishes Scala 3 artifacts :confetti_ball:

[All pull requests necessary](https://github.com/playframework/playframework/issues?q=label%3Ascala3+is%3Aclosed) were labeled with https://github.com/playframework/playframework/labels/scala3

Everyone is welcome to try it and give feedback. You can also get started with a seed project by using `sbt new playframework/play-scala-seed.g8 --branch 2.9.x` (or using `play-java-seed.g8` for our Java folks). Don't forget to switch to Scala `3.3.0-RC3` in `build.sbt` or when prompted.

Please report any problems you experience. There will be a Scala 3 migration guide where we will document what to look after and issues to expect: https://github.com/playframework/playframework/pull/11739

Now that we publish Scala 3 artifacts it does make sense to wait with the Play 2.9 release until Scala 3.3.0 final gets released, since it's that will be [a Scala LTS release](https://www.scala-lang.org/blog/2022/08/17/long-term-compatibility-plans.html#our-solution-scala-lts--scala-next) and [smart people advice to not use Scala 3.0/3.1/3.2 at this point anymore](https://github.com/sbt/sbt/discussions/7194#discussioncomment-5462505).
Shortly after Scala 3.3.0 there will be Scala 2.13.11 and 2.12.18, which won't happen before April 20th according to these comments:
* https://contributors.scala-lang.org/t/scala-2-13-11-release-planning/6088/4
* https://contributors.scala-lang.org/t/scala-2-12-18-release-planning/6089/5
* https://contributors.scala-lang.org/t/3-3-0-release-thread/6079

These Scala releases will also unblock just released Java 20, so even though not officially supported (only 11 LTS and 17 LTS  will be), I will give it a shot and run all tests with it, just to check future Java compatibility (https://github.com/coursier/jvm-index/pull/186 needed for that :)

So for Play that means the 2.9 release will be postponed shortly after the above Scala releases. Until that you can expect release candidates published with migration docs updated. I will update [the roadmap](https://github.com/orgs/playframework/projects/3) as well shortly (I didn't really stick to it since the Scala 3 migration completely sucked me in... ).
